gcc/rust/checks/errors/borrowck/rust-function-collector.h

```diff
@@ -0,0 +1,198 @@
+// Copyright (C) 2020-2023 Free Software Foundation, Inc.
+
+// This file is part of GCC.
+
+// GCC is free software; you can redistribute it and/or modify it under
+// the terms of the GNU General Public License as published by the Free
+// Software Foundation; either version 3, or (at your option) any later
+// version.
+
+// GCC is distributed in the hope that it will be useful, but WITHOUT ANY
+// WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+// for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with GCC; see the file COPYING3.  If not see
+// <http://www.gnu.org/licenses/>.
+
+#ifndef RUST_HIR_FUNCTION_COLLECTOR_H
+#define RUST_HIR_FUNCTION_COLLECTOR_H
+
+#include "rust-hir-item.h"
+#include "rust-hir-visitor.h"
+#include "rust-hir.h"
+
+#include <vector>
+
+namespace Rust {
+
+// Collects all HIR items eligible for borrow checking.
+class FunctionCollector : public HIR::HIRFullVisitor
+{
+  std::vector<HIR::Function *> functions;
+  std::vector<HIR::ClosureExpr *> closures;
+
+public:
+  void go (HIR::Crate &crate) { visit_all (crate.get_items ()); }
+
+  [[nodiscard]] const std::vector<HIR::Function *> &get_functions () const
+  {
+    return functions;
+  }
+
+  [[nodiscard]] const std::vector<HIR::ClosureExpr *> &get_closures () const
+  {
+    return closures;
+  }
+
+protected:
+  template <typename T> void visit_all (std::vector<std::unique_ptr<T>> &items)
+  {
+    for (std::unique_ptr<T> &item : items)
+      {
+	item->accept_vis (*this);
+      }
+  }
```

Comment:
IIUC you also need to call this on all the statements of a block, and all the items of a module right? this can be done in a later PR

Reply:
Yes, as mentioned bellow, that is a TODO.